### DISCUSSION TITLE
Possibility of multicore parallel rendering of bitmap?

### DISCUSSION THREAD

**Discussion by sandric** [Bodmer/TFT_eSPI#2530] [General]
Hi, I have rp2040 MCU and SSD1963 7 inch screen module working in 8 bit parallel mode.

During my tests with rp2040 overclocked to 225 MHz clock speeds - to render 750KB 16-bit 800*480 bitmap straight from flash takes around 150ms.

So I'm wondering if this rate is bounded by rp2040 MCU speeds or SSD1963 tft driver? And if its in any way possible to parallelize rendering by two cores - say render top half on one and bottom on another?

I saw some examples with teensy 4.1 with huge 600MHz clock speeds and from rendering speeds it looks almost video-capable like. Or it is bound more by RAM and buffer size mb? Very interesting to hear some thought of how to make it a bit more performant.

Thx.

**Comment by Bodmer** [comment]
That is slow, it should be faster, there could be a number of reasons for the slowness. The interface speed is not the limiting factor in this case.
Maybe it is the transfer approach used in your tests or another reason.

**Reply by sandric** [reply]
Thanks for reply! My code example - slightly modified your example of `TFT_Flash_Bitmap`, I just added my own 2 images that are C-encoded 800*480 16 bit images of bridge and a lake, made with this [tool](https://notisrac.github.io/FileToCArray/) . I specifically wanted to test raw speed from flash, without SD cards SPI/SDIO overhead etc, so not sure what can cause delay, mb only wires since its on a breadboard:

```C++
#include <TFT_eSPI.h>       // Hardware-specific library
#include "Mine.h"

TFT_eSPI tft = TFT_eSPI();  // Invoke custom library

void setup()
{
  Serial.begin(115200);

  delay(2000);

  
  tft.begin();
  tft.setRotation(1);
  tft.fillScreen(TFT_BLACK);
  tft.setSwapBytes(true);
  tft.pushImage(0, 0, mineWidth, mineHeight, mine_bridge);

  delay(2000);
}

void loop()
{
  long lTime;

  lTime = micros();
  
  tft.pushImage(0, 0, mineWidth, mineHeight, mine_bridge);
  
  lTime = micros() - lTime;
  Serial.printf("First in %d us\n", (int)lTime);

  delay(1000);

  lTime = micros();

  tft.pushImage(0, 0, mineWidth, mineHeight, mine_lake);

  lTime = micros() - lTime;
  Serial.printf("Second in %d us\n", (int)lTime);

  delay(1000);
}
```

`Mine.h` is in this format, I don't post it here because its quite big:
```C++
const uint16_t mineWidth = 800;
const uint16_t mineHeight = 480;

// array size is 768000
static const unsigned short mine_bridge[] PROGMEM  = {
  0x5141, 0x5141, 0x5141, 0x5141, 0x5141, 0x5141, 0x5162,
  ...
};

// array size is 768000
static const unsigned short mine_lake[] PROGMEM  = {
  0x9b87, 0xe5af, 0xedcf, 0xedce, 0xedce, 0xedce, 0xedae,
  ...
};
```

And here's what I see in serial logs:

```
Second in 158414 us
First in 158375 us
Second in 158534 us
First in 158450 us
Second in 158358 us
First in 158425 us
Second in 158412 us
First in 158506 us
Second in 158487 us
First in 158428 us
Second in 158474 us
First in 158445 us
```

My `UserSetup.h` if needed:

```C++
#define USER_SETUP_ID 303
#define TFT_PARALLEL_8_BIT
#define SSD1963_800ALT_DRIVER
#define TFT_RGB_ORDER TFT_BGR

#define TFT_DC   15
#define TFT_WR   14
#define TFT_RST  13

#define TFT_D0   2
#define TFT_D1   3
#define TFT_D2   4
#define TFT_D3   5
#define TFT_D4   6
#define TFT_D5   7
#define TFT_D6   8
#define TFT_D7   9

#define LOAD_GLCD
#define LOAD_FONT2
#define LOAD_FONT4
#define LOAD_FONT6
#define LOAD_FONT7
#define LOAD_FONT8
#define LOAD_GFXFF
#define SMOOTH_FONT
```

What is your guess normal speed should be?

**Reply by Bodmer** [reply]
I don't have access to testing hardware at the moment. It looks like the FLASH acessis slowing down the rendering. I will investigate but not for a few weeks.

**Reply by sandric** [reply]
Sure, thanks. In the meantime I'll experiment a bit with parallel execution of rendering with both cores by splitting bitmap into two images.
